tools/smithy-rs-sync/src/fs.rs

```diff
@@ -0,0 +1,308 @@
+
+    handwritten_files
+        .run_callback_on_generated_files(&directory, callback)
+        .context(here!())
+}
+
+pub fn find_handwritten_files_and_folders(
+    aws_sdk_path: &Path,
+    build_artifacts_path: &Path,
+) -> anyhow::Result<Vec<PathBuf>> {
+    let mut paths = Vec::new();
+    let callback = |path: &Path| -> Result<(), std::io::Error> {
+        paths.push(path.to_owned());
+        Ok(())
+    };
+    let dotfile_path = aws_sdk_path.join(HANDWRITTEN_DOTFILE);
+    let handwritten_files = HandwrittenFiles::from_dotfile(&dotfile_path).context(here!())?;
+
+    handwritten_files
+        .run_callback_on_handwritten_files(&build_artifacts_path, callback)
+        .context(here!())?;
+
+    Ok(paths)
+}
+
+/// A struct with methods that help when checking to see if a file is handwritten or
+/// automatically generated.
+///
+/// # Examples
+///
+/// for a .handwritten_files that looks like this:
+///
+/// ```
+/// # Only paths referring to siblings of the .handwritten_files file are valid
+///
+/// # this file will be protected
+/// file.txt
+/// # this folder will be protected
+/// folder1/
+/// ```
+///
+/// ```rust
+///   let handwritten_files = HandwrittenFiles::from_dotfile(
+///       Path::new("/Users/zelda/project/.handwritten_files")
+///     ).unwrap();
+///
+///   assert!(handwritten_files.is_handwritten(Path::new("file.txt")));
+///   assert!(handwritten_files.is_handwritten(Path::new("folder1/")));
+/// }
+/// ```
+#[derive(Debug)]
+pub struct HandwrittenFiles<'file> {
+    files_and_folders: gitignore::File<'file>,
+}
+
+impl<'file> HandwrittenFiles<'file> {
+    pub fn from_dotfile(dotfile_path: &'file Path) -> Result<Self, HandwrittenFilesError> {
+        let files_and_folders = gitignore::File::new(dotfile_path)?;
+
+        Ok(Self { files_and_folders })
+    }
+
+    pub fn is_handwritten(&self, path: &'file Path) -> Result<bool, HandwrittenFilesError> {
+        self.files_and_folders.is_excluded(path).map_err(Into::into)
+    }
+
+    pub fn run_callback_on_generated_files(
+        &self,
+        directory: &Path,
+        mut callback: impl FnMut(&Path) -> Result<(), std::io::Error>,
+    ) -> Result<(), HandwrittenFilesError> {
+        let read_dir = std::fs::read_dir(directory)?;
+
+        for entry in read_dir {
+            let path = &entry.map(|e| e.path())?;
+
+            let is_generated = !self.is_handwritten(path)?;
+            if is_generated {
+                callback(path)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn run_callback_on_handwritten_files(
+        &self,
+        directory: &Path,
+        mut callback: impl FnMut(&Path) -> Result<(), std::io::Error>,
+    ) -> Result<(), HandwrittenFilesError> {
+        let read_dir = std::fs::read_dir(directory)?;
+
+        for entry in read_dir {
+            let path = &entry.map(|e| e.path())?;
+
+            if self.is_handwritten(path)? {
+                callback(path)?;
+            }
+        }
```

Comment:
this is fine. One option could be to restructure it a bit. Instead of having this use callbacks, you could use `impl Iterator` instead, something like `read_dir.filter(|...|)`. Then you could have `handwritten_files(&self) -> impl Iterator<Item=PathBuf>` which might be a little cleaner to use elsewhere

Reply:
The thing that makes this difficult is the `Result`s that need to be propagated. I refactored this to use the scan iterator instead but it's kinda gross and still requires a `collect` because of the lifetime of `gitignore::File` prevents me from easily passing an iterator with an internal ref to `HandwrittenFiles`. I'm sure it's possible, I just didn't want to get too sucked in.